[Shen Yi Hong] iP

shenyih0ng opened the pull request:
# Duke

> “⭐️⭐️⭐️⭐️⭐️” - average app store rating

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

## Installation

All you need to do is,

1. Download it from [here](https://github.com/shenyih0ng/ip/releases/tag/v0.1.0).
2. Run `duke.jar`
    - `java -jar duke.jar`
3. Add your tasks.
4. Let it manage your tasks for you 🎉
5. It is that **simple**!

## Features:

- [X] Managing tasks
- [X] Managing deadlines
- [X] Managing events
- [X] Tag your tasks (e.g. `#school`, `#cs2103`)

---

If you are a `Java` programmer, you can use it to practice `Java` too. Here's the main method:

```java
public static void main(String[] args) {
    Application.launch(Main.class, DEFAULT_APPLICATION_NAME, DEFAULT_CACHE_PATH);
}
```

> https://github.com/shenyih0ng/ip/blob/c1255503947c956cfc8df00bce032cbd0338b7a4/src/main/java/duke/Launcher.java#L5-L12

shenyih0ng committed:
[feat] Added support for echo and terminate

- add utility functions to format responses

shenyih0ng committed:
[feat] Added support for `list` command

- max capacity of Duke's list/history is set to 100

shenyih0ng committed:
Added Command enums and Task class

shenyih0ng committed:
[feat] Added support for `mark`, `unmark` commands

- added error handing for `mark`, `unmark` commands
- improve semantics
  - `history` -> `taskList`

shenyih0ng committed:
Added Task(`Todo`, `Deadline`, `Event`) classes

- refactor base `Task` class to be abstract

shenyih0ng committed:
[feat] Support for `todo`, `deadline`, `event` commands

- update available commands
  - added a `UNKNOWN` command to handle command errors

shenyih0ng committed:
Refactor structure, componentize `CommandHandlers`

- added `utils` and `data` packages

shenyih0ng committed:
Integrate new structure

- simplify errors with simple placeholder outputs

shenyih0ng committed:
Rename `commandHandler` package to `command`

shenyih0ng committed:
[bug fix] Return error msg when `mark/unmark 0`

shenyih0ng committed:
[tests] Added text ui tests

shenyih0ng committed:
Add `CommandException` and refactor error handling

shenyih0ng committed:
[struct] Use exceptions to handle command errors

shenyih0ng committed:
[test] Update text ui test

shenyih0ng committed:
[test] Add more tests for `mark` command

- tests for invalid task idx inputs

shenyih0ng committed:
[feat] Add support for `delete` command

shenyih0ng committed:
[test] Add text ui tests for `delete` command

shenyih0ng committed:
[chore] Set up minimal `gradle` build pipeline

shenyih0ng committed:
[feat] Add simple serialize/deseralize ability

- currently Duke saves to cache on each command

shenyih0ng committed:
Refactor `save` to depend on command

shenyih0ng committed:
[refactor] Use regex for command checking

- simplify `CommandHandler` interface
- refactor `CommandHandler.run` arguments

shenyih0ng committed:
[test] Update auto text ui test

shenyih0ng committed:
Replace manual command checking with `regex.Matcher` (#2)

shenyih0ng committed:
[bugfix] Fix ui loop to adhere to #2 changes

shenyih0ng committed:
Merge branch 'master' into feat/branch-Level-7

shenyih0ng committed:
[feat] Add `datetime` parsing and validation

shenyih0ng committed:
[test] Add tests for `datetime`

shenyih0ng committed:
[feat] Add more details to error message

- refactor Duke response logic and structure

shenyih0ng committed:
[test] Update test with new error messages

shenyih0ng committed:
Add support for saving and loading `TaskList` (#1)

shenyih0ng committed:
Merge branch 'master' into feat/branch-Level-8

shenyih0ng committed:
Support for `datetime` parsing and validation (#3)

shenyih0ng committed:
[refactor] Improve OOP structure

- squash commit from `refactor/oop` branch

shenyih0ng committed:
[chore] Shift classes into new package structure

- update `gradle` run main class to `duke.Duke`
- update `text-ui-test` to compile java files under `src/main/java/duke/`

shenyih0ng committed:
[refactor] Shift validation to `CommandHandler`

shenyih0ng committed:
[bugfix] Fix `deadline` command expected output

- update `runtest` scripts to respect new package structure

shenyih0ng committed:
[test] Add unit test for `CommandFactory`

- add gradle task for unit tests

shenyih0ng committed:
[test] Add unit test for `CommandDeadlineHandler`

shenyih0ng committed:
[chore] Add `shadowJar` gradle task

- add `testLogging` to gradle test

shenyih0ng committed:
[doc] Add docs for command factory and handlers

shenyih0ng committed:
[doc] Add docs for task response classes

shenyih0ng committed:
[doc] Add docs for task classes

shenyih0ng committed:
[doc] Add docs for `Duke` and `Storage`

shenyih0ng committed:
[chore] Add `checkstyle` task to gradle

shenyih0ng committed:
[chore] Fix `checkstyle` violations

shenyih0ng committed:
[doc] Add `JavaDoc`

shenyih0ng committed:
[chore] Add `checkstyle` config files

shenyih0ng committed:
[chore] Add `checkstyle` and fix violations

shenyih0ng committed:
[refactor] Shift `TaskList` display to `toString`

- update `CommandListHandler`

shenyih0ng committed:
[feat] Add support for `find` command

shenyih0ng committed:
[test] Add text-ui tests for `find` command

shenyih0ng committed:
[feat] Handle empty `TaskList` for `find` command

- add more tests

shenyih0ng committed:
Support for `find` command (#4)

ryanguai commented on src/main/java/duke/data/tasks/TaskTodo.java:
```diff
@@ -0,0 +1,19 @@
+package duke.data.tasks;
+
+public class TaskTodo extends Task {
+    private static final long serialVersionUID = 21L;
```
If UID is an acronym, you can consider changing it to lower case as per the coding standard.

ryanguai commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,13 @@
+package duke.command;
+
+public enum Command {
+    BYE, // termination
+    TODO, // add Todo task
+    DEADLINE, // add Deadline task
+    EVENT, // add Event task
+    LIST, // display list
+    MARK, // mark a task
+    UNMARK, // unmark a task
+    DELETE, // delete a task
+    FIND, // find tasks
+}
```
Excellent use of enums.

ryanguai commented on src/main/java/duke/command/CommandException.java:
```diff
@@ -0,0 +1,7 @@
+package duke.command;
+
+public class CommandException extends Exception {
+    public CommandException(String errorMsg) {
```
Perhaps you can write out "Message" in full? I feel that "Msg" may not be universally understood.

ryanguai commented on src/main/java/duke/command/CommandFactory.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    public CommandHandler getCommandHandler(Command command, String commandStr)
+            throws CommandException {
+        switch (command) {
+        case LIST:
+            return new CommandListHandler(commandStr);
+        case MARK:
+        case UNMARK:
+            return new CommandMarkHandler(commandStr);
+        case TODO:
+            return new CommandTodoHandler(commandStr);
+        case DEADLINE:
+            return new CommandDeadlineHandler(commandStr);
+        case EVENT:
+            return new CommandEventHandler(commandStr);
+        case DELETE:
+            return new CommandDeleteHandler(commandStr);
+        case FIND:
+            return new CommandFindHandler(commandStr);
+        case BYE:
+            return new CommandTerminateHandler(commandStr);
+        default:
+            return null;
+        }
```
Well done for avoiding indentation for your case clauses!

ryanguai commented on src/main/java/duke/command/response/CommandResponse.java:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * Constructor for CommandResponse class
+     *
+     * @param responseStr response string
+     * @param triggerSave boolean to trigger a save to cache
+     * @param triggerTerminate boolean to trigger a termination
```
You can add a punctuation (full stop) behind your param descriptions.

ryanguai commented on src/test/java/duke/command/handler/CommandDeadlineHandlerTest.java:
```diff
@@ -0,0 +1,52 @@
+        for (String normalInput : normalInputs) {
+            try {
+                new CommandDeadlineHandler(normalInput);
+            } catch (CommandException ignored) {
```
Perhaps you can name the exception as "error" instead? I feel it would make more sense to use a noun.

ryanguai commented on src/main/java/duke/data/tasks/Task.java:
```diff
@@ -0,0 +1,42 @@
+public abstract class Task implements Serializable {
+
+    protected final String title;
+    protected boolean marked;
```
You can change the name to sound more like a boolean eg. isMarked.

ryanguai reviewed (commented):
Overall, your code is written mostly in accordance with the coding standard. Well done! Aside from the comments I wrote, you can also add in more documentation for the remaining methods and for each class.

shenyih0ng committed:
[chore] Add `JavaFX` dependencies to gradle

shenyih0ng committed:
[feat] Add simple `GUI` skeleton

- deprecated `text-ui` functions

shenyih0ng committed:
Add duke and user display images

shenyih0ng committed:
[feat] Add custom `dialogBox` for user and duke

- some aesthetic changes to the GUI

shenyih0ng committed:
[refactor] Add `fxml` and remove `Ui` class

shenyih0ng committed:
[refactor] Simplify `getCommandHandler` api

shenyih0ng committed:
[refactor] Migrate UI logic to `.fxml`

shenyih0ng committed:
[feat] Add termination delay

shenyih0ng committed:
[chore] Fix `checkstyle` violation

shenyih0ng committed:
[feat] Add GUI

- deprecate `text-ui`

maryjess commented on src/main/java/duke/command/CommandFactory.java:
```diff
@@ -0,0 +1,62 @@
+public class CommandFactory {
+
+    /**
+     * Parse command given by user to a predefined command enum
```
Perhaps can change `parse` to `parses` here?

maryjess commented on src/main/java/duke/command/handler/CommandDeleteHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+public class CommandDeleteHandler extends CommandHandler {
+
+    protected static final String INVALID_FORMAT_MSG = String.join("\n",
```
I like how you structure the invalid format message!

maryjess commented on src/main/java/duke/command/handler/CommandMarkHandler.java:
```diff
@@ -0,0 +1,65 @@
+            } else {
+                task.unmark();
+            }
+            return new UpdateTaskResponse(task, toMark ? UpdateType.MARK : UpdateType.UNMARK);
```
Nice use of enum here

maryjess reviewed (commented):
Overally, neat and clean code! 👍
A common recurring issue I saw is that some of the Javadoc comments do not use simple present tense.

shenyih0ng committed:
temp: Suppress `JavaDoc` checks

shenyih0ng committed:
[ci] Add Java CI github action

shenyih0ng committed:
Fix typo

shenyih0ng committed:
Remove `text-ui` tests in CI

shenyih0ng committed:
Add CI status badge to README

shenyih0ng committed:
Add Java CI github action (#5)

shenyih0ng committed:
[chore] Change gradle `mainClassName` to GUI Launcher

shenyih0ng committed:
[chore] Enable assertions in gradle run

shenyih0ng committed:
[refactor] Add assertions

shenyih0ng committed:
[refactor] Change `msg` abbreviation to `message`

shenyih0ng committed:
[refactor] Change `marked` to `isMarked`

shenyih0ng committed:
[refactor] Change invocation of `parseCommand`

- Directly invoke from `CommandFactory` instead of a instance

shenyih0ng committed:
[refactor] Modify `DEFAULT_CACHE_PATH` constant to `String`

shenyih0ng committed:
Add assertions (#6)

shenyih0ng committed:
Refactor for Code Quality (#7)

shenyih0ng committed:
[refactor] Add `CommandAddTaskHandler` abstraction

shenyih0ng committed:
[chore] Fix `checkstyle` violation

shenyih0ng committed:
[feat] Add support for tagging Tasks

shenyih0ng committed:
[refactor] Modify `findTasks` to more general `filterTasks`

shenyih0ng committed:
[feat] Add support for `list #tag`

shenyih0ng committed:
[feat] Add support for `tag` command

shenyih0ng committed:
[refactor] Add `UpdateTaskHandler` abstraction

- package base command handler into `command.handler.base`

shenyih0ng committed:
[test] Add unit tests for `tag` command

shenyih0ng committed:
[feat] Add support for task tagging (#8)

Right now there is no easy way to group tasks or ascribe a category to a task.

Allowing task tagging will greatly improve user experience as they are able to group similar tasks together with a tag e.g. `#CS2103`/`#admin`

Let's,
- allow users to specify tags anywhere in the title of a task (inline tags) and we will automatically extract and process these tags
- give users the option to also use a `tag` command to tag existing tasks without a inline tag (e.g. `tag #admin 1`)
- provide a way to list only tasks under a specific tag using the `list` command (e.g. `list #CS2103`)

shenyih0ng committed:
[refactor] Better handling of empty `TaskList`

- let `TaskList` handle display instead of each `Command`

shenyih0ng committed:
[feat] Auto resizeable GUI

- colored `DialogBox` border

shenyih0ng committed:
[ui] Use different font for `Duke` response

shenyih0ng committed:
Add warning emoji to error message

- smiley face to welcome user :)

shenyih0ng committed:
Fix boolean naming and header comments

shenyih0ng committed:
More JavaDoc fixes

shenyih0ng committed:
GUI Improvements (#10)

- Auto resizeable window
- Used a different font to differentiate between Duke and user
- Colored border for better UX

shenyih0ng committed:
Code quality improvements (#11)

shenyih0ng committed:
Merge branch 'master' into branch-A-CodingStandard

shenyih0ng committed:
Minor fixes

shenyih0ng committed:
[cleanup] Merge unmerged branches (`branch-A-CodingStandard`) (#12)

shenyih0ng committed:
Merge branch 'master' into branch-Level-10

shenyih0ng committed:
cleanup: Merge unmerged branches (branch-Level-10) (#13)

shenyih0ng committed:
Merge branch 'master' into branch-A-JavaDoc

shenyih0ng committed:
Minor fixes

shenyih0ng committed:
[cleanup] Merge unmerged branches (`branch-A-JavaDoc`) (#14)

shenyih0ng committed:
Add product screenshot

shenyih0ng committed:
cleanup: Merge squashed branch (`branch-Level-10`) (#15)

shenyih0ng committed:
cleanup: Merge squashed branch (`branch-A-JavaDoc`) (#16)

shenyih0ng committed:
Add user guide

shenyih0ng committed:
[bugfix] Handle duplicate task tags

shenyih0ng committed:
Add user guide (#17)
